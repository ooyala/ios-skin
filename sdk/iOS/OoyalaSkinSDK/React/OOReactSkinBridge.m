//
//  OOReactSkinBridge.m
//  OoyalaSkinSDK
//
//  Created on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridge.h"
#import "OOReactSkinBridgeModule.h"
#import <React/RCTBridge+Private.h>

@interface OOReactSkinBridge ()

@property (nonatomic, weak, readonly) id<OOReactSkinBridgeDelegate> delegate;

@end


@implementation OOReactSkinBridge

@dynamic delegate;

- (instancetype)initWithDelegate:(id<OOReactSkinBridgeDelegate>)delegate
                   launchOptions:(NSDictionary *)launchOptions {
  return [super initWithDelegate:delegate launchOptions:launchOptions];
}

- (void)invalidate {
  [super invalidate];
  [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (void)setUp {
  [super setUp];
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(didInitializeModuleNotification:)
                                             name:RCTDidInitializeModuleNotification
                                           object:self.batchedBridge];
}

- (void)didInitializeModuleNotification:(NSNotification *)notification {
  id module = notification.userInfo[@"module"];

  if ([self.delegate respondsToSelector:@selector(bridge:didLoadModule:)]) {
    if (module && [module conformsToProtocol:@protocol(OOReactSkinBridgeModule)]) {
      [self.delegate bridge:self didLoadModule:module];
    }
  }
}

@end


@implementation RCTBridge (OOReactSkinEventsEmitterable)

- (OOReactSkinEventsEmitter *)skinEventsEmitter {
  return [self moduleForName:[OOReactSkinEventsEmitter moduleName]];
}

@end
