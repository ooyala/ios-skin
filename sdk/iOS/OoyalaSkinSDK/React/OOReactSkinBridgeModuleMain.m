//
//  OOReactSkinBridgeModuleMain.m
//  OoyalaSkinSDK
//
//  Created by Maksim Kupetskii on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridgeModuleMain.h"
#import "NSMutableDictionary+Utils.h"

#import <OoyalaSDK/OODebugMode.h>
#import <OoyalaSDK/OOOoyalaPlayer.h>

@interface OOReactSkinBridgeModuleMain ()

@property (nonatomic, weak, nullable) id<OOReactSkinModelDelegate> skinModelDelegate;

@end


@implementation OOReactSkinBridgeModuleMain

@synthesize bridge;

RCT_EXPORT_MODULE(OOReactSkinBridgeModuleMain);

- (dispatch_queue_t)methodQueue {
  return dispatch_get_main_queue();
}

- (OOReactSkinBridgeModuleType)moduleType {
  return OOReactSkinBridgeModuleTypeMain;
}

- (void)setSkinViewDelegate:(nonnull id<OOReactSkinModelDelegate>)delegate {
  _skinModelDelegate = delegate;
}

#pragma mark - Constants
#pragma mark Buttons
static NSString *playPauseButtonName    = @"PlayPause";
static NSString *playButtonName         = @"Play";
static NSString *rewindButtonName       = @"rewind";
static NSString *socialShareButtonName  = @"SocialShare";
static NSString *fullscreenButtonName   = @"fullscreen";
static NSString *learnMoreButtonName    = @"LearnMore";
static NSString *skipButtonName         = @"Skip";
static NSString *moreOptionButtonName   = @"More";
static NSString *upNextDismiss          = @"upNextDismiss";
static NSString *upNextClick            = @"upNextClick";
static NSString *adIconButtonName       = @"Icon";
static NSString *adOverlayButtonName    = @"Overlay";
static NSString *stereoscopicButtonName = @"stereoscopic";
static NSString *pipButtonName          = @"PIP";
static NSString *replayButtonName       = @"replay";

#pragma mark Keys
static NSString *nameKey              = @"name";
static NSString *embedCodeKey         = @"embedCode";
static NSString *percentageKey        = @"percentage";
static NSString *languageKey          = @"language";
static NSString *bucketInfoKey        = @"bucketInfo";
static NSString *actionKey            = @"action";
static NSString *audioTrackKey        = @"audioTrack";
static NSString *indexKey             = @"index";
static NSString *clickUrlKey          = @"clickUrl";
static NSString *eventNameKey         = @"eventName";
static NSString *playbackSpeedRateKey = @"playbackSpeedRate";
static NSString *volumeKey            = @"volume";

#pragma mark Values
static NSString *startValue   = @"start";
static NSString *moveValue    = @"move";
static NSString *endValue     = @"end";
static NSString *clickValue   = @"click";
static NSString *impressValue = @"impress";

#pragma mark - Exported Methods

RCT_EXPORT_METHOD(onVolumeChanged:(NSDictionary *)parameters) {
  float volume = [parameters[volumeKey] floatValue];
  [self.skinModelDelegate handleVolumeChanged:volume];
}

RCT_EXPORT_METHOD(onMounted) {
  LOG(@"onMounted - Not going to use at the moment");
}

RCT_EXPORT_METHOD(onLanguageSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleLanguageSelection:parameters[languageKey]];
}

RCT_EXPORT_METHOD(onAudioTrackSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleAudioTrackSelection:parameters[audioTrackKey]];
}

RCT_EXPORT_METHOD(onPlaybackSpeedRateSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handlePlaybackSpeedRateSelection:parameters[playbackSpeedRateKey]];
}

RCT_EXPORT_METHOD(onPress:(NSDictionary *)parameters) {
  NSString *buttonName = parameters[nameKey];

  if ([buttonName isEqualToString:playPauseButtonName]) {
    [self.skinModelDelegate handlePlayPause];
  } else if([buttonName isEqualToString:playButtonName]) {
    [self.skinModelDelegate handlePlay];
  } else if([buttonName isEqualToString:replayButtonName]) {
    [self.skinModelDelegate handleReplay];
  } else if([buttonName isEqualToString:rewindButtonName]) {
    [self.skinModelDelegate handleRewind];
  } else if([buttonName isEqualToString:socialShareButtonName]) {
    [self.skinModelDelegate handleSocialShare];
  } else if([buttonName isEqualToString:fullscreenButtonName]) {
    [self.skinModelDelegate toggleFullscreen];
  } else if([buttonName isEqualToString:learnMoreButtonName]) {
    [self.skinModelDelegate handleLearnMore];
  } else if([buttonName isEqualToString:skipButtonName]) {
    [self.skinModelDelegate handleSkip];
  } else if ([buttonName isEqualToString:adIconButtonName]) {
    [self.skinModelDelegate handleIconClick:[parameters[indexKey] integerValue]];
  } else if([buttonName isEqualToString:moreOptionButtonName]) {
    [self.skinModelDelegate handleMoreOption];
  } else if([buttonName isEqualToString:upNextDismiss]) {
    [self.skinModelDelegate handleUpNextDismiss];
  } else if([buttonName isEqualToString:upNextClick]) {
    [self.skinModelDelegate handleUpNextClick];
  } else if ([buttonName isEqualToString:adOverlayButtonName]) {
    [self.skinModelDelegate handleOverlay:parameters[clickUrlKey]];
  } else if ([buttonName isEqualToString:pipButtonName]) {
    [self.skinModelDelegate handlePip];
  } else if ([buttonName isEqualToString:stereoscopicButtonName]) {
    [self.skinModelDelegate toggleStereoMode];
  }
}

RCT_EXPORT_METHOD(handleTouchStart:(NSDictionary *)params){
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: startValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchMove:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: moveValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchEnd:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: endValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(onScrub:(NSDictionary *)parameters) {
  Float64 position = [parameters[percentageKey] doubleValue];
  [self.skinModelDelegate handleScrub:position];
}

RCT_EXPORT_METHOD(setEmbedCode:(NSDictionary *)parameters) {
  NSString *embedCode = parameters[embedCodeKey];
  [self.skinModelDelegate setEmbedCode:embedCode];
}

RCT_EXPORT_METHOD(onDiscoveryRow:(NSDictionary *)parameters) {
  NSString *action     = parameters[actionKey];
  NSString *bucketInfo = parameters[bucketInfoKey];

  if ([action isEqualToString:clickValue]) {
    NSString *embedCode = parameters[embedCodeKey];
    [self.skinModelDelegate handleDiscoveryClick:bucketInfo embedCode:embedCode];
  } else if ([action isEqualToString:impressValue]) {
    [self.skinModelDelegate handleDiscoveryImpress:bucketInfo];
  }
}

@end
