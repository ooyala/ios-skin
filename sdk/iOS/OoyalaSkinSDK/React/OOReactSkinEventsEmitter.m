/**
 * @file       OOReactSkinEventsEmitter.m
 * @class      OOReactSkinEventsEmitter OOReactSkinEventsEmitter.m "OOReactSkinEventsEmitter.m"
 * @brief      OOReactSkinEventsEmitter
 * @details    OOReactSkinEventsEmitter.m in OoyalaSDK
 * @date       8/13/18
 * @copyright Copyright (c) 2018 Ooyala, Inc. All rights reserved.
 */

#import "OOReactSkinEventsEmitter.h"
#import "OOConstant.h"
#import "OOQueuedEvent.h"

#import <OoyalaSDK/OOOoyalaPlayer.h>
#import <OoyalaSDK/OOClosedCaptions.h>
#import <OoyalaSDK/OODebugMode.h>

@interface OOReactSkinEventsEmitter ()

@property (atomic) NSMutableArray *queuedEvents; //QueuedEvent *

@end

@implementation OOReactSkinEventsEmitter

RCT_EXPORT_MODULE(OOReactSkinEventsEmitter);

+ (BOOL)requiresMainQueueSetup {
  return YES;
}

- (instancetype)init {
  if (self = [super init]) {
    _isReactReady = NO;
    _queuedEvents = [NSMutableArray new];
  }
  return self;
}

- (NSArray<NSString *> *)supportedEvents {
  return @[@"timeChanged",
           @"seekStarted",
           @"seekCompleted",
           @"ccStylingChanged",
           @"currentItemChanged",
           @"frameChanged",
           @"fullscreenToggled",
           @"volumeChanged",
           @"playCompleted",
           @"stateChanged",
           @"desiredStateChanged",
           @"discoveryResultsReceived",
           @"onClosedCaptionUpdate",
           @"adStarted",
           @"adSwitched",
           @"adPodStarted",
           @"adPodCompleted",
           @"adOverlay",
           @"adError",
           @"setNextVideo",
           @"upNextDismissed",
           @"playStarted",
           @"error",
           @"embedCodeSet",
           @"controllerKeyPressEvent",
           @"vrContentEvent",
           @"multiAudioEnabled",
           @"audioTrackChanged",
           @"playbackSpeedEnabled",
           @"playbackSpeedRateChanged"];
}

- (void)sendDeviceEventWithName:(NSString *)eventName body:(id)body {
  if (self.isReactReady) {
    if (![eventName isEqualToString:OOOoyalaPlayerTimeChangedNotification] &&
        ![eventName isEqualToString:OO_CLOSED_CAPTIONS_UPDATE_EVENT]) {
      LOG(@"sendDeviceEventWithName: %@", eventName);
    }
    [self sendEventWithName:eventName body:body];
  } else {
    [self queueEventWithName:eventName body:body];
  }
}

- (void)setIsReactReady:(BOOL)isReactReady {
  if (_isReactReady) {
    LOG(@"received ReactReady notification after ready");
    return;
  }
  _isReactReady = isReactReady;
  // PurgeEvents must happen after isReactReady, however, I'm not positive this is truly thread-safe.
  // If a notification is queued during PurgeEvents, there could be an execption
  [self purgeEvents];
}

- (void)queueEventWithName:(NSString *)eventName body:(id)body {
  LOG(@"Queued Event: %@", eventName);
  OOQueuedEvent *event = [[OOQueuedEvent alloc] initWithWithName:eventName body:body];
  [self.queuedEvents addObject:event];
}

- (void)purgeEvents {
  LOG(@"Purging Events to skin");
  for (OOQueuedEvent *event in self.queuedEvents) {
    [self sendDeviceEventWithName:event.eventName body:event.body];
  }
  [self.queuedEvents removeAllObjects];
}

@end
