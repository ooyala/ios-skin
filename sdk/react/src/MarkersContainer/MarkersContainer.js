// @flow

import * as React from 'react';
import { StyleSheet, View } from 'react-native';

import Marker, { type MarkerType } from '../Marker';
import styles from './MarkersContainer.styles';

type Props = {
  accentColor?: string,
  duration: number,
  markers: Array<MarkerType>,
  style?: StyleSheet.Styles,
};

const MarkersContainer = ({
  accentColor, duration, markers, style,
}: Props) => (
  markers.length > 0
    ? (
      <View style={[styles.root, style]}>
        {markers.map((marker, index) => (
          <Marker
            accentColor={accentColor}
            duration={duration}
            key={index} // eslint-disable-line react/no-array-index-key
            marker={marker}
          />
        ))}
      </View>
    )
    : null
);

MarkersContainer.defaultProps = {
  accentColor: undefined,
  style: undefined,
};

export default MarkersContainer;
