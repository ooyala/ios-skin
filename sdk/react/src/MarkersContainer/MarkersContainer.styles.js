// @flow

import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  root: {
    height: 0,
    left: 0,
    position: 'absolute',
    top: 0,
    width: '100%',
  },
});
