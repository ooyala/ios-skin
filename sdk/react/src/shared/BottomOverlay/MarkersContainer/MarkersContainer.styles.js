// @flow

import { StyleSheet } from 'react-native';

import { MARKERS_SIZES } from '../../../constants';

export default StyleSheet.create({
  root: {
    height: MARKERS_SIZES.CONTAINER_HEIGHT,
  },
});
