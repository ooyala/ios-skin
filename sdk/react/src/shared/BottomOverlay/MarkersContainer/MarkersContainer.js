// @flow

import React from 'react';
import { Animated } from 'react-native';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

import Marker from './Marker';
import styles from './MarkersContainer.styles';
import type { Marker as MarkerType } from '../../../types/Markers';

type Props = {
  accentColor?: ?string,
  duration: number,
  markers: Array<MarkerType>,
  onSeek: number => void,
  style?: ViewStyleProp,
  width: number,
};

type State = {
  lastTouchedIndex?: number,
};

export default class MarkersContainer extends React.Component<Props, State> {
  static defaultProps = {
    accentColor: undefined,
    style: undefined,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      lastTouchedIndex: undefined,
    };
  }

  handleTouch(index: number) {
    this.setState({ lastTouchedIndex: index });
  }

  render() {
    const {
      accentColor, duration, markers, onSeek, style, width,
    } = this.props;
    const { lastTouchedIndex } = this.state;

    if (markers.length === 0) {
      return null;
    }

    return (
      // Path pointer events through the View to its children.
      <Animated.View pointerEvents="box-none" style={[styles.root, { width }, style]}>
        {markers.map((marker, index) => (
          <Marker
            accentColor={accentColor}
            containerWidth={width}
            duration={duration}
            key={index} // eslint-disable-line react/no-array-index-key
            marker={marker}
            onSeek={onSeek}
            onTouch={() => this.handleTouch(index)}
            style={{
              // If the marker has been touched, then raise it over the others, otherwise the former should overlap the
              // latter.
              zIndex: lastTouchedIndex === index ? markers.length : markers.length - index,
            }}
          />
        ))}
      </Animated.View>
    );
  }
}
