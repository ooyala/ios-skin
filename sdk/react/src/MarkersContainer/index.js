// @flow

import MarkersContainer from './MarkersContainer';

export default MarkersContainer;
