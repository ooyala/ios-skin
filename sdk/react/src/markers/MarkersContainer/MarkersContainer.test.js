// @flow

import React from 'react';
import TestRenderer from 'react-test-renderer';

import MarkersContainer from './MarkersContainer';

describe('MarkersContainer', () => {
  it('renders matching snapshot', () => {
    const wrapper = TestRenderer.create(
      <MarkersContainer
        duration={60}
        markers={[
          {
            start: 'start',
            type: 'text',
            text: 'Hello, world!',
          },
          {
            iconUrl: 'http://example.com/icon.png',
            start: 'start',
            type: 'icon',
          },
        ]}
        onSeek={() => undefined}
      />,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('does not render anything when markers array is empty', () => {
    const wrapper = TestRenderer.create(
      <MarkersContainer
        duration={60}
        markers={[]}
        onSeek={() => undefined}
      />,
    );

    expect(wrapper.toJSON()).toBeNull();
  });
});
