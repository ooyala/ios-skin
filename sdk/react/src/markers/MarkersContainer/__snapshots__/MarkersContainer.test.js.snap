// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkersContainer renders matching snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "height": 64,
    }
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "black",
          "borderColor": "#4E4E4E",
          "borderRadius": 2,
          "borderWidth": 1,
          "bottom": 6,
          "padding": 2,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": -35,
            },
          ],
          "width": 70,
        },
        Object {
          "left": "0%",
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      numberOfLines={4}
      style={
        Object {
          "color": "white",
          "fontSize": 13,
          "lineHeight": 13,
          "textAlign": "center",
        }
      }
      suppressHighlighting={true}
    >
      Hello...
    </Text>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 0,
            "borderColor": "transparent",
            "borderTopColor": "#4E4E4E",
            "borderWidth": 6,
            "bottom": -6,
            "height": 0,
            "left": "50%",
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": -3,
              },
            ],
            "width": 0,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "black",
          "borderColor": "#4E4E4E",
          "borderRadius": 2,
          "borderWidth": 1,
          "bottom": 6,
          "height": 32,
          "padding": 2,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": -16,
            },
          ],
          "width": 32,
        },
        Object {
          "left": "0%",
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "http://example.com/icon.png",
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 0,
            "borderColor": "transparent",
            "borderTopColor": "#4E4E4E",
            "borderWidth": 6,
            "bottom": -6,
            "height": 0,
            "left": "50%",
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": -3,
              },
            ],
            "width": 0,
          },
          false,
        ]
      }
    />
  </View>
</View>
`;
